### DISCUSSION TITLE
Does it work auto pipeline when I use adapter like go-redis?

### DISCUSSION THREAD

**Discussion by ghost** [redis/rueidis#206] [Q&A] [answered]
I know there are adapter like go-redis (https://github.com/rueian/rueidis/tree/main/rueidiscompat)
But I'm not sure does it work auto pipelined when I use this adapter.

For example, when I use go-redis library, I use this code.
```
        pipe := client.Pipeline()

	for _, key := range keys {
		pipe.Expire(ctx, key, exp)
	}

	_, err := pipe.Exec(ctx)
	if err != nil {
	        ...
```

Howerver, I don't know how to write code like I wrote above when using rueidis adapter.
Is this code right when using adapter?
```
for _, key := range keys {
		err := client.Expire(ctx, key, exp).Err()
		if err != nil {
			...
		}
	}
```

**Comment by rueian** [comment, marked as answer]
Hi @KwonsooPark,

rueidis is able to pipeline concurrent requests automatically.

In your example, however, those `client.Expire(ctx, key, exp).Err()` aren't fired concurrently, and therefore they won't be pipelined.

Unfortunately, `client.Pipeline()` equivalent is not yet implemented in rueidiscompat. Would you consider using `DoMulti()` instead?

**Reply by ghost** [reply]
Thanks for your answer. I'll try to work with your suggestion :)

**Reply by t-kruth** [reply]
@rueian, curious about this answer as we're in the process of porting over from go-redisv9 to rueidis.  We have a timing hook to report redis round trip latency and noticed there's a change in latency when leveraging rueidis and the compat adapter.

if we use the compat package.. are we not gaining the pipelining by default?

**Reply by rueian** [reply]
Hi @t-kruth,

You should get auto pipelining by default. There is no way to skip auto pipelining in rueidis as long as you make concurrent requests on it.

How is your latency going? And do you change any default settings of rueidis?

**Reply by t-kruth** [reply]
thanks for getting back @rueian! Since it wasn't the initial thought we had, I created a separate discussion to not piggy back on this one...

https://github.com/redis/rueidis/discussions/609
